### PR TITLE
Enable plugin resolution as subcommand for selected builtin commands

### PR DESCRIPTION
**Pull request by ardaguclu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR adds external plugin resolution as subcommand for selected builtin commands if subcommand does not exist as builtin.

In it's alpha stage, this will only be enabled for create command and this feature is hidden behind `KUBECTL_ENABLE_CMD_SHADOW` environment variable.

#### Does this PR introduce a user-facing change?
```release-note
Enable external plugins can be used as subcommands for kubectl create command if subcommand does not exist as builtin only when KUBECTL_ENABLE_CMD_SHADOW environment variable is exported.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/3638-kubectl-plugin-subcommands
```

### CODE REVIEW

**Comment by ardaguclu** [comment]
/triage accepted
/priority important-longterm
/hold
for review

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd.go [hunk 1]
```diff
@@ -132,11 +136,39 @@ func NewDefaultKubectlCommandWithArgs(o KubectlOptions) *cobra.Command {
 			case "help", cobra.ShellCompRequestCmd, cobra.ShellCompNoDescRequestCmd:
 				// Don't search for a plugin
 			default:
-				if err := HandlePluginCommand(o.PluginHandler, cmdPathPieces); err != nil {
+				if err := HandlePluginCommand(o.PluginHandler, cmdPathPieces, true); err != nil {
 					fmt.Fprintf(o.IOStreams.ErrOut, "Error: %v\n", err)
 					os.Exit(1)
 				}
 			}
+		} else {
```
Comment (sttts) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:144]: nit: `else if`

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd.go [hunk 2]
```diff
@@ -244,6 +276,12 @@ func HandlePluginCommand(pluginHandler PluginHandler, cmdArgs []string) error {
 	for len(remainingArgs) > 0 {
 		path, found := pluginHandler.Lookup(strings.Join(remainingArgs, "-"))
 		if !found {
+			if !greedy {
+				// if greed is false, we shouldn't continue searching with shorter names.
```
Comment (sttts) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:280]:
```suggestion
				// if greedy is false, we shouldn't continue searching with shorter names.
```

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd.go [hunk 3]
```diff
@@ -224,7 +256,7 @@ func (h *DefaultPluginHandler) Execute(executablePath string, cmdArgs, environme
 
 // HandlePluginCommand receives a pluginHandler and command-line arguments and attempts to find
 // a plugin executable on the PATH that satisfies the given arguments.
-func HandlePluginCommand(pluginHandler PluginHandler, cmdArgs []string) error {
+func HandlePluginCommand(pluginHandler PluginHandler, cmdArgs []string, greedy bool) error {
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:259]: needed a moment to understand why this is greedy. Not sure that's a good term. How about "exactMatch" or "allowPrefixMatch" ?

[the remark this answers is not in the document]
Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:259]: `exactMatch` is indeed better naming than greedy. It will work opposite.

**Comment by ardaguclu** [comment]
Failures are not related to the changes in this PR;
/test pull-kubernetes-conformance-kind-ga-only-parallel
/test pull-kubernetes-integration

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd.go [hunk 4]
```diff
@@ -83,6 +83,10 @@ import (
 
 const kubectlCmdHeaders = "KUBECTL_COMMAND_HEADERS"
 
+var (
+	allowedCmdsSubcommandPlugin = map[string]struct{}{"create": {}}
```
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:87]: Nit: `sets.Set[string]` will be more appropriate and easier to work with by using `Has` etc.

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd.go [hunk 5]
```diff
@@ -132,11 +136,39 @@ func NewDefaultKubectlCommandWithArgs(o KubectlOptions) *cobra.Command {
...
 				// Don't search for a plugin
 			default:
-				if err := HandlePluginCommand(o.PluginHandler, cmdPathPieces); err != nil {
+				if err := HandlePluginCommand(o.PluginHandler, cmdPathPieces, false); err != nil {
 					fmt.Fprintf(o.IOStreams.ErrOut, "Error: %v\n", err)
 					os.Exit(1)
 				}
 			}
+		} else if err == nil {
+			if cmdutil.CmdPluginAsSubcommand.IsEnabled() {
```
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:145]: Nit: what I think Stefan meant with that `else if` to make these two lines:
```golang
else if cmdutil.CmdPluginAsSubcommand.IsEnabled()
```

**Review by soltysh — APPROVED** [review]
/lgtm
/hold cancel

**Comment by sftim** [comment]
(IMO) we should mention this in changelogs, even if the new behavior is semi-hidden. It's still _there_.

**Comment by sftim** [comment]
Thanks @ardaguclu

**Pull request merged**